x-pack/packages/ml/aiops_log_rate_analysis/queries/fetch_log_rate_analysis_for_alert.ts

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { queue } from 'async';
+import { chunk } from 'lodash';
+
+import type * as estypes from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+
+import { withSpan } from '@kbn/apm-utils';
+import type { ElasticsearchClient } from '@kbn/core/server';
+import type { SignificantItem } from '@kbn/ml-agg-utils';
+import { getSampleProbability } from '@kbn/ml-random-sampler-utils';
+
+import type { AiopsLogRateAnalysisSchema } from '../api/schema';
+import { getLogRateAnalysisParametersFromAlert } from '../get_log_rate_analysis_parameters_from_alert';
+import { getSwappedWindowParameters } from '../get_swapped_window_parameters';
+import { getLogRateChange } from '../get_log_rate_change';
+import { getBaselineAndDeviationRates } from '../get_baseline_and_deviation_rates';
+import { getLogRateAnalysisTypeForCounts } from '../get_log_rate_analysis_type_for_counts';
+import { LOG_RATE_ANALYSIS_TYPE } from '../log_rate_analysis_type';
+
+import { fetchIndexInfo } from './fetch_index_info';
+import { fetchSignificantCategories } from './fetch_significant_categories';
+import { fetchSignificantTermPValues } from './fetch_significant_term_p_values';
+
+const MAX_CONCURRENT_QUERIES = 5;
+const CHUNK_SIZE = 50;
+
+interface QueueItem {
+  fn: typeof fetchSignificantCategories | typeof fetchSignificantTermPValues;
+  fieldNames: string[];
+}
+
+/**
+ * Runs log rate analysis data on an index given some alert metadata.
+ */
+export async function fetchLogRateAnalysisForAlert({
+  esClient,
+  abortSignal,
+  arguments: args,
+}: {
+  esClient: ElasticsearchClient;
+  abortSignal?: AbortSignal;
+  arguments: {
+    index: string;
+    alertStartedAt: string;
+    alertEndedAt?: string;
+    alertRuleParameterTimeSize?: number;
+    alertRuleParameterTimeUnit?: string;
+    timefield?: string;
+    searchQuery?: estypes.QueryDslQueryContainer;
+  };
+}) {
+  const { alertStartedAt, alertEndedAt, timefield = '@timestamp' } = args;
+
+  const { timeRange, windowParameters } = getLogRateAnalysisParametersFromAlert({
+    alertStartedAt,
+    alertEndedAt,
```

Comment:
From a glance this leads me to believe that the time range that is analysed is from the alert starts until the alert ends. Looking at the implementation I see that's not quite the case. So already here it's pretty opaque. 

That being said, why do we even care about the alert end time? Afaict we should focus on the events leading up to the alert. Meaning the end timestamp for the analysed time range should be `alertStartedAt`.

Reply:
We embed the log rate analysis UI on the same page further down below, so in this PR I tried to replicate the same baseline+deviation selection for the analysis, my first thoughts were we shouldn't do a different analysis under the hood for this contextual insight.

I annotated in red in this screenshot what's going on how we do the selection:

<img width="1689" alt="image" src="https://github.com/user-attachments/assets/f0acd9ee-44d5-44c1-8fd0-acf753a48a96">

For the embedded UI we consider the alert end to possibly capture the whole time range the logs were above the threshold and not just leading up to it. We adjust the deviation selection to consider the lookback durations to ensure we 1) start the deviation one lookback duration before the alert start to capture logs leading up to the alert start and 2) end the deviation one lookback before the alert ended to ignore the time range where logs were already again below the threshold.

For comparison I added in purple the 15m time range `getLogCategories()` picks for the contextual insight. 

Looking at this and considering your suggestion we could update the baseline+deviation similar to the yellow annotation I did in the screenshot: The deviation would be just one lookback period leading up to the alert start, in this case 5 minutes. The baseline would be the 10 minutes before the deviation. We could then ignore the alert end. I wouldn't hardcode these 5/10 minute intervals though, but scale them according to the lookback interval like it's done now. Otherwise we might not pick a good deviation time range for customized lookback periods.  So for a 1m lookback interval, the deviation duration would be 1 minute too and the baseline would start 3 minutes before the alert start. The deviation would always be equal to the lookback duration and the baseline would always start 3 lookback durations before the alert start.

Because we don't expose the raw data of the analysis anyway, I think it would be fine to not have the exact same baseline+deviation compared to the embedded UI. Instead it will be more consistent within this contextual insight and all the different analysis will consider data up to the alert start like you suggested.

What do you think?

Reply:
I went ahead and updated the PR in [5fa2a43](https://github.com/elastic/kibana/pull/187690/commits/5fa2a43d69ac4d2f624600a3cfcba675f759ecfd) to include the suggestion from above, it now doesn't need the alert end anymore.

Reply:
Sounds good, thanks!